server/src/main/scala/manifests/graph/BatchStage.scala

```diff
@@ -58,43 +85,54 @@ class BatchStage(now: () => SDateLike) extends GraphStage[FlowShape[List[Arrival
+        push(outArrivals, lookupBatch.toList)
       } else log.info(s"Nothing to push right now")
     }
 
-    private def updatePrioritisedAndSubscribers(existingSubscribers: Map[ArrivalKey, Option[MillisSinceEpoch]], existingPrioritised: Set[ArrivalKey]): Set[ArrivalKey] = {
-      log.info(s"about to check all arrivals for those due a lookup")
-      val updatedPrioritised: Set[ArrivalKey] = addToPrioritised(existingSubscribers, existingPrioritised)
+    private def pushRegisteredArrivalsUpdates(): Unit = if (registeredArrivalsUpdates.nonEmpty) {
+      log.info(s"Pushing ${registeredArrivalsUpdates.size} registered arrivals updates")
+      push(outRegisteredArrivals, RegisteredArrivals(registeredArrivalsUpdates))
+      registeredArrivalsUpdates = Map()
+    }
+
+    private def purgeExpired(): Unit = {
+      lookupQueue = Crunch.purgeExpired(lookupQueue, (a: ArrivalKey) => a.scheduled, now, expireAfterMillis)
+      registeredArrivals = Crunch.purgeExpiredTuple(registeredArrivals, (a: ArrivalKey) => a.scheduled, now, expireAfterMillis)
+    }
 
-      val (prioritisedBatch, remainingPrioritised) = updatedPrioritised.splitAt(500)
+    private def updatePrioritisedAndSubscribers(): Set[ArrivalKey] = {
+      val currentNow = now()
+      val updatedLookupQueue: Set[ArrivalKey] = refreshLookupQueue(currentNow)
 
-      log.info(s"prioritisedBatch: ${prioritisedBatch.size}. remainingPrioritised: ${remainingPrioritised.size}")
-      lookupQueue = remainingPrioritised
+      val (nextLookupBatch, remainingLookups) = updatedLookupQueue.toSeq.sortBy(_.scheduled).splitAt(batchSize)
+
+      lookupQueue = Set[ArrivalKey](remainingLookups: _*)
 
       val lookupTime: MillisSinceEpoch = now().millisSinceEpoch
 
-      registeredArrivals = prioritisedBatch.foldLeft(registeredArrivals) {
-        case (subscribersSoFar, priorityArrival) => subscribersSoFar.updated(priorityArrival, Option(lookupTime))
+      nextLookupBatch.foreach { arrivalForLookup =>
+        registeredArrivals = registeredArrivals.updated(arrivalForLookup, Option(lookupTime))
+        registeredArrivalsUpdates = registeredArrivalsUpdates.updated(arrivalForLookup, Option(lookupTime))
       }
 
-      prioritisedBatch
+      Set[ArrivalKey](nextLookupBatch: _*)
     }
 
-    private def addToPrioritised(subscribersToCheck: Map[ArrivalKey, Option[MillisSinceEpoch]], existingPrioritised: Set[ArrivalKey]): Set[ArrivalKey] = {
-      subscribersToCheck.foldLeft(existingPrioritised) {
-        case (prioritisedSoFar, (subscriber, None)) =>
-          if (!prioritisedSoFar.contains(subscriber)) prioritisedSoFar + subscriber
+    private def refreshLookupQueue(currentNow: SDateLike): Set[ArrivalKey] = registeredArrivals
+      .foldLeft(lookupQueue) {
+        case (prioritisedSoFar, (arrival, None)) =>
+          if (!prioritisedSoFar.contains(arrival)) prioritisedSoFar + arrival
```

Comment:
I think this would be clearer if you put a line break after the condition.

Reply:
done